### PR TITLE
Reenable 4244

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
This reenables 4244 in all places other than `mono/`

Contributes to https://github.com/dotnet/runtime/issues/66154

I broke down the updates by general area.

@Maoni0 @PeterSolMS See the GC commit
@janvorli There is a single exception walk logic update. See the Debugger commit
@BruceForstall There is a single SuperPMI update - See the SuperPMI commit

@jkotas @davidwrighton @elinor-fung Eyes on the rest are most welcome.

@am11 @janvorli I updated zlib-intel. Is there an upstream I can create a PR for?

### CODE REVIEW

File: src/coreclr/debug/daccess/daccess.cpp [hunk 1]
```diff
@@ -4766,12 +4766,15 @@ ClrDataAccess::SetAllTypeNotifications(
...
 ClrDataAccess::SetAllCodeNotifications(
     /* [in] */ IXCLRDataModule* mod,
-    /* [in] */ ULONG32 flags)
+    /* [in] */ ULONG32 flagsRaw)
 {
     HRESULT status;
 
     DAC_ENTER();
 
+    _ASSERTE(flagsRaw <= USHRT_MAX);
```
Comment (jkotas) [inline src/coreclr/debug/daccess/daccess.cpp:4775]: This assert is not generally valid. It is valid only after `IsValidMethodCodeNotification` has been called.

File: src/coreclr/vm/amd64/cgencpu.h [hunk 2]
```diff
@@ -141,10 +141,10 @@ void     R8ToFPSpill(void* pSpillSlot, SIZE_T  srcDoubleAsSIZE_T)
 // Parameter size
 //**********************************************************************
 
-inline unsigned StackElemSize(unsigned parmSize, bool isValueType = false /* unused */, bool isFloatHfa = false /* unused */)
+inline uint16_t StackElemSize(unsigned parmSize, bool isValueType = false /* unused */, bool isFloatHfa = false /* unused */)
```
Comment (jkotas) [inline src/coreclr/vm/amd64/cgencpu.h:144]: I would keep it as unsigned and fix the callsite that needs it in smaller type to do the cast.

File: src/coreclr/vm/callconvbuilder.cpp [hunk 3]
```diff
@@ -320,7 +320,7 @@ HRESULT CallConv::TryGetUnmanagedCallingConventionFromModOpt(
     _In_ PCCOR_SIGNATURE pSig,
     _In_ ULONG cSig,
     _Inout_ CallConvBuilder* builder,
-    _Out_ UINT *errorResID)
+    _Out_ WORD *errorResID)
```
Comment (jkotas) [inline src/coreclr/vm/callconvbuilder.cpp:323]: We use `UINT` for resource IDs. For example, `RealCOMPlusThrowHR` that takes resource IDs has it as UINT. Why is this change necessary?

**Comment by am11** [comment]
> @am11 @janvorli I updated zlib-intel. Is there an upstream I can create a PR for?

Upstream info of external libs are in their version files: https://github.com/dotnet/runtime/blob/694067c12626668074197d2f85f7babe6f01ac7a/src/native/external/zlib-intel-version.txt#L3

Note that @lambdageek is in the process of updating zlib and zlib-intel: https://github.com/dotnet/runtime/pull/68219.

File: src/coreclr/vm/amd64/cgencpu.h (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/amd64/cgencpu.h:144]: I thought about that. However, this is a really weird API to have such loose value for. The concern I have is weird casting for a size function. I don't like casting away sizes because that breeds trouble. This approach ensures we know the size it what we want it to be. Does that convince you in this case?

File: src/coreclr/vm/callconvbuilder.cpp (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/callconvbuilder.cpp:323]: Sad. This could have been done in a better way. Sorry. I will revert this.

File: src/coreclr/vm/methodtablebuilder.h [hunk 4]
```diff
@@ -2073,7 +2073,7 @@ class MethodTableBuilder
 
     // --------------------------------------------------------------------------------------------
     // Used for analyzing overlapped fields defined by explicit layout types.
-    enum bmtFieldLayoutTag {empty, nonoref, oref, byref};
+    enum bmtFieldLayoutTag : BYTE {empty, nonoref, oref, byref};
```
Comment (jkotas) [inline src/coreclr/vm/methodtablebuilder.h:2076]: It would be also nice to change the type of `BYTE *pFieldLayout` and similar argument of related methods to `bmtFieldLayoutTag*`.

File: src/coreclr/vm/amd64/cgencpu.h (hunk 2 above)
Comment (jkotas) [reply src/coreclr/vm/amd64/cgencpu.h:144]: Where are we casting away the size for this one?

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/amd64/cgencpu.h:144]: Keeping this `unsigned` requires a cast at

https://github.com/dotnet/runtime/blob/ff00ac7a52225d1c528bccb174aae448542d49ac/src/coreclr/vm/mlinfo.cpp#L2976-L2985

Comment (janvorli) [reply src/coreclr/vm/amd64/cgencpu.h:144]: It is kind of strange that with this change the function gets unsigned parmSize and after aligning it returns a smaller type uint16_t.

File: src/coreclr/vm/comdelegate.cpp [hunk 5]
```diff
@@ -531,8 +531,8 @@ BOOL GenerateShuffleArrayPortable(MethodDesc* pMethodSrc, MethodDesc *pMethodDst
         {
             ShuffleEntry entry = (*pShuffleEntryArray)[i];
 
-            int srcIndex = GetNormalizedArgumentSlotIndex(entry.srcofs);
-            int dstIndex = GetNormalizedArgumentSlotIndex(entry.dstofs);
+            UINT16 srcIndex = GetNormalizedArgumentSlotIndex(entry.srcofs);
+            UINT16 dstIndex = GetNormalizedArgumentSlotIndex(entry.dstofs);
 
             _ASSERTE((srcIndex >= 0) && ((unsigned int)srcIndex < argSlots));
```
Comment (janvorli) [inline src/coreclr/vm/comdelegate.cpp:537]: It seems we don't need the (unsigned int) cast after your change.

File: src/coreclr/vm/methodtablebuilder.h (hunk 4 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/methodtablebuilder.h:2076]: Good idea. Done.

File: src/native/external/zlib-intel/deflate.c [hunk 6]
```diff
@@ -1484,10 +1484,10 @@ local block_state deflate_stored(s, flush)
         _tr_stored_block(s, (char *)0, 0L, last);
 
         /* Replace the lengths in the dummy stored block with len. */
-        s->pending_buf[s->pending - 4] = len;
-        s->pending_buf[s->pending - 3] = len >> 8;
-        s->pending_buf[s->pending - 2] = ~len;
-        s->pending_buf[s->pending - 1] = ~len >> 8;
+        s->pending_buf[s->pending - 4] = (Bytef)len;
```
Comment (janvorli) [inline src/native/external/zlib-intel/deflate.c:1487]: I wonder if we should touch sources of the external libraries unless we are planning to backport the change upstream. This could cause some pain if we update it in the future.

File: src/coreclr/debug/daccess/daccess.cpp (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/debug/daccess/daccess.cpp:4775]: I'm not sure why. All uses of the `flags` argument are implicitly down cast to `USHORT`.

File: src/coreclr/vm/amd64/cgencpu.h (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/amd64/cgencpu.h:144]: I've reverted it and added an assert.

File: src/native/external/zlib-intel/deflate.c (hunk 6 above)
Comment (AaronRobinsonMSFT) [reply src/native/external/zlib-intel/deflate.c:1487]: We've been doing this for libunwind too. I submitted an upstream PR: https://github.com/libunwind/libunwind/pull/333. I'm fine doing either, just let me know.

File: src/coreclr/debug/daccess/daccess.cpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/debug/daccess/daccess.cpp:4775]: This is public method that is presumably called by code outside this repo. It is trying to do argument validation and return `E_INVALIDARG` when the arguments are wrong. 

If the method is called with `flags: 0x10001`, it should return `E_INVALIDARG` and not assert.

File: src/native/external/zlib-intel/deflate.c (hunk 6 above)
Comment (AaronRobinsonMSFT) [reply src/native/external/zlib-intel/deflate.c:1487]: Based on https://github.com/dotnet/runtime/pull/68615#issuecomment-1111412946, I'm just going to revert these changes and suppress the warning in that project.

**Comment by AaronRobinsonMSFT** [comment]
@am11 and @janvorli I've updated the cmake files for zlib and zlib-intel to suppress 4244. This will likely need to be discussed with @GrabYourPitchforks at some point.

File: src/coreclr/debug/daccess/daccess.cpp (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/debug/daccess/daccess.cpp:4775]: I see. It won't do that right now because `IsValidMethodCodeNotification(flags)` implicitly downcasts to `USHORT`. I will update to do as suggested.

File: src/coreclr/tools/superpmi/superpmi-shared/methodcontextreader.cpp [hunk 7]
```diff
@@ -28,7 +28,7 @@ HANDLE MethodContextReader::OpenFile(const char* inputFile, DWORD flags)
 static std::string to_lower(const std::string& input)
 {
     std::string res = input;
-    std::transform(input.cbegin(), input.cend(), res.begin(), tolower);
+    std::transform(input.cbegin(), input.cend(), res.begin(), (char(*)(char))tolower);
```
Comment (BruceForstall) [inline src/coreclr/tools/superpmi/superpmi-shared/methodcontextreader.cpp:31]: It seems dangerous to lie about the function signature.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontextreader.cpp:31]: I can instead create a wrapper if desired.

**Comment by BruceForstall** [comment]
The (new) SuperPMI change LGTM.

**Comment by BruceForstall** [comment]
@TIHan PTAL at ilasm changes

**Review by hoyosjs — COMMENTED** [review]
Changes under src/coreclr/debug lgtm

**Comment by AaronRobinsonMSFT** [comment]
@elinor-fung Thanks. The following doesn't seem like it is anything I've done:

```
API Error attempting to talk to Docker Server:500 Server Error: Internal Server Error ("CreateComputeSystem b383d37b60e2cd59ad9126c53bb7e6946dc7f782171e4def1f060da2658f1555: The request is not supported.
```

**Comment by AaronRobinsonMSFT** [comment]
The timeout/failure above is tracked in https://github.com/dotnet/runtime/issues/67728.

File: src/coreclr/debug/daccess/daccess.cpp [hunk 8]
```diff
@@ -4981,14 +4981,15 @@ ClrDataAccess::SetCodeNotifications(
                 {
                     for (ULONG32 check = 0; check < numTokens; check++)
                     {
-                        if (!IsValidMethodCodeNotification(flags[check]))
+                        _ASSERTE(flags[check] <= USHRT_MAX);
```
Comment (jkotas) [inline src/coreclr/debug/daccess/daccess.cpp:4984]: This should not be needed. IsValidMethodCodeNotification was changed to take ULONG

**Review by jkotas — APPROVED** [review]
LGTM otherwise

**Comment by AaronRobinsonMSFT** [comment]
This seems to be a failure with uploading logs. The tests appear to have all passed.

/cc @dotnet/runtime-infrastructure 

```
Starting: Publish SuperPMI logs
==============================================================================
Task         : Publish Pipeline Artifacts
Description  : Publish (upload) a file or directory as a named artifact for the current run
Version      : 1.199.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/publish-pipeline-artifact
==============================================================================
Artifact name input: SuperPMI_Logs_x64_checked
Uploading pipeline artifact from D:\a\_work\1\s\artifacts\spmi_logs\ for build #1751868
Using default max parallelism.
Max dedup parallelism: 192
ApplicationInsightsTelemetrySender will correlate events with X-TFS-Session 99b3589b-74bb-4465-8404-9d202c2e4a5b
DedupManifestArtifactClient will correlate http requests with X-TFS-Session 99b3589b-74bb-4465-8404-9d202c2e4a5b
10 files processed.
Processed 10 files from D:\a\_work\1\s\artifacts\spmi_logs\ successfully.
Uploaded 245,954 out of 245,954 bytes
Content upload is done!

Content upload statistics:
Total Content: 0.5 MB
Physical Content Uploaded: 25.8 KB
Logical Content Uploaded: 0.2 MB
Compression Saved: 0.2 MB
Deduplication Saved: 0.2 MB
Number of Chunks Uploaded: 11
Total Number of Chunks: 21

ApplicationInsightsTelemetrySender correlated 2 events with X-TFS-Session 99b3589b-74bb-4465-8404-9d202c2e4a5b
##[error]Artifact SuperPMI_Logs_x64_checked already exists for build 1751868.
Finishing: Publish SuperPMI logs
```

```
Starting: Publish SuperPMI build logs
==============================================================================
Task         : Publish Pipeline Artifacts
Description  : Publish (upload) a file or directory as a named artifact for the current run
Version      : 1.199.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/publish-pipeline-artifact
==============================================================================
Artifact name input: SuperPMI_BuildLogs_x64_checked
Uploading pipeline artifact from D:\a\_work\1\s\artifacts\log for build #1751868
Using default max parallelism.
Max dedup parallelism: 192
ApplicationInsightsTelemetrySender will correlate events with X-TFS-Session 55a4c631-7e5b-442b-a442-7a8a42f76f37
DedupManifestArtifactClient will correlate http requests with X-TFS-Session 55a4c631-7e5b-442b-a442-7a8a42f76f37
3 files processed.
Processed 3 files from D:\a\_work\1\s\artifacts\log successfully.
Uploaded 538,189 out of 597,510 bytes
Content upload is done!

Content upload statistics:
Total Content: 1.2 MB
Physical Content Uploaded: 0.5 MB
Logical Content Uploaded: 0.5 MB
Compression Saved: 57.0 bytes
Deduplication Saved: 0.7 MB
Number of Chunks Uploaded: 9
Total Number of Chunks: 27

ApplicationInsightsTelemetrySender correlated 2 events with X-TFS-Session 55a4c631-7e5b-442b-a442-7a8a42f76f37
##[error]Artifact SuperPMI_BuildLogs_x64_checked already exists for build 1751868.
Finishing: Publish SuperPMI build logs
```
